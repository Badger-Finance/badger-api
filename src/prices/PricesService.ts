import { Service } from '@tsed/common';
import { InternalServerError } from '@tsed/exceptions';
import NodeCache from 'node-cache';
import fetch from 'node-fetch';
import { PriceData, PriceSummary, TokenPrice } from '../interface/TokenPrice';
import { COINGECKO_URL, TOKENS } from '../util/constants';
import { getSushiswapPrice, getUniswapPrice } from '../util/util';

const priceCache = new NodeCache({ stdTTL: 300, checkperiod: 480 });

/**
 * API price oracle service. Uses CoinGecko as a source of truth for most
 * tokens when possible, and TheGraph for AMM pairs when not available
 * via CG. Prices are cached for 5 minutes at a time, but may live up to 8.
 */
@Service()
export class PriceService {
	/**
	 * Retrieve the USD price for a given token balance.
	 * @param contract Token contract address.
	 * @param balance Token balance to calculate price.
	 */
	async getUsdValue(contract: string, balance: number): Promise<number> {
		const tokenPrice = await this.getTokenPriceData(contract);
		return tokenPrice.usd * balance;
	}

	/**
	 * Retrieve the ETH price for a given token balance.
	 * @param contract Token contract address.
	 * @param balance Token balance to calculate price.
	 */
	async getEthValue(contract: string, balance: number): Promise<number> {
		const tokenPrice = await this.getTokenPriceData(contract);
		return tokenPrice.eth * balance;
	}

	/**
	 * Retrieve the price data for a given token in USD and ETH.
	 * @param contract Token contract address.
	 */
	async getTokenPriceData(contract: string): Promise<TokenPrice> {
		const cachedPrice: TokenPrice | undefined = priceCache.get(contract);
		if (!cachedPrice) {
			const priceData = await this.getPriceData();
			return priceData[contract];
		}
		return cachedPrice;
	}

	/**
	 * Retrieve all protocol token prices in both USD and ETH.
	 */
	async getPriceData(): Promise<PriceData> {
		const [
			badgerPrice,
			diggPrice,
			wbtcPrice,
			uniBadgerWbtcPrice,
			uniDiggWbtcPrice,
			slpBadgerWbtcPrice,
			slpDiggWbtcPrice,
			slpEthWbtcPrice,
			renBtcPrice,
			sBtcPrice,
			tBtcPrice,
			wethPrice,
			sushiPrice,
		] = await Promise.all([
			getContractPrice(TOKENS.BADGER),
			getContractPrice(TOKENS.DIGG),
			getContractPrice(TOKENS.WBTC),
			getUniswapPrice(TOKENS.UNI_BADGER_WBTC),
			getUniswapPrice(TOKENS.UNI_DIGG_WBTC),
			getSushiswapPrice(TOKENS.SUSHI_BADGER_WBTC),
			getSushiswapPrice(TOKENS.SUSHI_DIGG_WBTC),
			getSushiswapPrice(TOKENS.SUSHI_ETH_WBTC),
			getContractPrice(TOKENS.CRV_RENBTC),
			getTokenPrice('sbtc'),
			getTokenPrice('tbtc'),
			getContractPrice(TOKENS.WETH),
			getContractPrice(TOKENS.SUSHI),
		]);
		const priceData: PriceData = {};
		priceData[TOKENS.BADGER] = badgerPrice;
		priceData[TOKENS.DIGG] = diggPrice;
		priceData[TOKENS.WBTC] = wbtcPrice;
		priceData[TOKENS.UNI_BADGER_WBTC] = uniBadgerWbtcPrice;
		priceData[TOKENS.UNI_DIGG_WBTC] = uniDiggWbtcPrice;
		priceData[TOKENS.SUSHI_BADGER_WBTC] = slpBadgerWbtcPrice;
		priceData[TOKENS.SUSHI_DIGG_WBTC] = slpDiggWbtcPrice;
		priceData[TOKENS.SUSHI_ETH_WBTC] = slpEthWbtcPrice;
		priceData[TOKENS.CRV_RENBTC] = renBtcPrice;
		priceData[TOKENS.CRV_SBTC] = sBtcPrice;
		priceData[TOKENS.CRV_TBTC] = tBtcPrice;
		priceData[TOKENS.WETH] = wethPrice;
		priceData[TOKENS.SUSHI] = sushiPrice;
		return priceData;
	}

	async getPriceSummary(): Promise<PriceSummary> {
		const priceData = await this.getPriceData();
		const priceSummary: PriceSummary = {};
		for (const [key, value] of Object.entries(priceData)) {
			priceSummary[key] = value.usd;
		}
		return priceSummary;
	}
}

/**
 * Retrieve the price data for a given token in USD and ETH.
 * Warning: Not all contracts are supported.
 * If a token is not supported, but available on CoinGecko, use getTokenPrice.
 * @param contract Token contract address.
 * @throws {InternalServerError} Failed price lookup.
 */
export const getContractPrice = async (contract: string): Promise<TokenPrice> => {
	const cachedPrice: TokenPrice | undefined = priceCache.get(contract);
	if (cachedPrice) return cachedPrice;
	const response = await fetch(
		`${COINGECKO_URL}/token_price/ethereum?contract_addresses=${contract}&vs_currencies=usd,eth`,
	);
	if (!response.ok) throw new InternalServerError('Unable to query contract price');
	const json = await response.json();
	const contractKey = contract.toLowerCase(); // coingecko return key in lower case
	if (!json[contractKey] || !json[contractKey].usd || !json[contractKey].eth)
		throw new InternalServerError('Unable to resolve contract price');
	const contractPrice: TokenPrice = {
		address: contract,
		usd: json[contractKey].usd,
		eth: json[contractKey].eth,
	};
	priceCache.set(contract, contractPrice);
	return contractPrice;
};

/**
 * Retrieve the price data for a given token in USD and ETH.
 * @param token CoinGecko token name.
 * @throws {InternalServerError} Failed price lookup.
 */
export const getTokenPrice = async (token: string): Promise<TokenPrice> => {
	const cachedPrice: TokenPrice | undefined = priceCache.get(token);
	if (cachedPrice) return cachedPrice;
	const response = await fetch(`${COINGECKO_URL}/price?ids=${token}&vs_currencies=usd,eth`);
	if (!response.ok) throw new InternalServerError('Unable to query token price');
	const json = await response.json();
	if (!json[token] || !json[token].usd || !json[token].eth)
		throw new InternalServerError('Unable to resolve token price');
	const tokenPrice: TokenPrice = {
		name: token,
		usd: json[token].usd,
		eth: json[token].eth,
	};
	priceCache.set(token, tokenPrice);
	return tokenPrice;
};
