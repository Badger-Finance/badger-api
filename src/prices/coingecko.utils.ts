import { InternalServerError } from '@tsed/exceptions';
import { Chain } from '../chains/config/chain.config';
import { COINGECKO_URL } from '../config/constants';
// TODO: generalize and add some axios utilities
import { request } from '../etherscan/etherscan.utils';
import { PriceData } from '../tokens/interfaces/price-data.interface';
import { getToken, getTokenByName } from '../tokens/tokens.utils';
import { CoinGeckoPriceResponse } from './interface/coingecko-price-response.interface';
import { TokenPrice } from './interface/token-price.interface';

export async function fetchPrices(chain: Chain, inputs: string[], lookupName = false): Promise<PriceData> {
  let baseURL;
  let params: Record<string, string>;

  // utilize coingecko name look up api
  if (lookupName) {
    baseURL = `${COINGECKO_URL}/price`;
    params = {
      ids: inputs.join(','),
      vs_currencies: 'usd',
    };
  } else {
    baseURL = `${COINGECKO_URL}/token_price/${chain.network}`;
    params = {
      contract_addresses: inputs.join(','),
      vs_currencies: 'usd',
    };
  }

  const expectedTokens: string[] = [];
  const result = await request<CoinGeckoPriceResponse>(baseURL, params);
  const priceData = Object.fromEntries(
    Object.entries(result).map((entry) => {
      const [key, value] = entry;
      let token;
      try {
        token = getToken(key);
      } catch {
        token = getTokenByName(key);
      }
      const { address } = token;
      expectedTokens.push(address);
      return [address, { address, price: value.usd }];
    }),
  );
  const missingTokens: string[] = [];
  expectedTokens.forEach((input) => {
    if (!priceData[input]) {
      missingTokens.push(input);
    }
  });
  if (missingTokens.length > 0) {
    console.error(`Missing prices for:\n${missingTokens.join('\n')}`);
  }
  return priceData;
}

/**
 * Retrieve the price data for a given token in USD and ETH.
 * Warning: Not all contracts are supported.
 * If a token is not supported, but available on CoinGecko, use getTokenPrice.
 * @param contract Token contract address.
 * @throws {InternalServerError} Failed price lookup.
 */
export async function getContractPrice(contract: string): Promise<TokenPrice> {
  const params = {
    contract_addresses: contract,
    vs_currencies: 'usd',
  };
  const result = await request<Record<string, { usd: number }>>(`${COINGECKO_URL}/token_price/ethereum`, params);
  const contractKey = contract.toLowerCase(); // coingecko return key in lower case
  if (!result[contractKey] || !result[contractKey].usd) {
    throw new InternalServerError(`Unable to resolve ${contract} price by contract`);
  }
  const token = getToken(contract);
  return {
    address: token.address,
    price: result[contractKey].usd,
  };
}

/**
 * Retrieve the price data for a given token in USD and ETH.
 * @param token CoinGecko token name.
 * @throws {InternalServerError} Failed price lookup.
 */
export async function getTokenPrice(name: string): Promise<TokenPrice> {
  const params = {
    ids: name,
    vs_currencies: 'usd',
  };
  const result = await request<Record<string, { usd: number }>>(`${COINGECKO_URL}/price`, params);
  if (!result[name] || !result[name].usd) {
    throw new InternalServerError(`Unable to resolve ${name} price by name`);
  }
  const token = getTokenByName(name);
  return {
    address: token.address,
    price: result[name].usd,
  };
}
