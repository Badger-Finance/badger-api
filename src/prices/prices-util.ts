import { BadRequest, InternalServerError, NotFound, UnprocessableEntity } from '@tsed/exceptions';
import { DocumentClient, QueryInput } from 'aws-sdk/clients/dynamodb';
import { ethers } from 'ethers';
import NodeCache = require('node-cache');
import { getItems, saveItem } from '../aws/dynamodb-utils';
import { ChainStrategy } from '../chains/strategies/chain.strategy';
import { COINGECKO_URL, PRICE_DATA, TOKENS } from '../config/constants';
import { Token } from '../tokens/interfaces/token.interface';
import { PriceData, TokenPrice, TokenPriceSnapshot } from '../tokens/interfaces/token-price.interface';
import { getToken, getTokenByName, protocolTokens } from '../tokens/tokens-util';
import { TokenConfig } from '../tokens/types/token-config.type';
import AttributeValue = DocumentClient.AttributeValue;
import fetch from 'node-fetch';
import { loadChains } from '../chains/chain';
import { Chain } from '../chains/config/chain.config';
import { getSett } from '../setts/setts-util';
import { TokenType } from '../tokens/enums/token-type.enum';

/**
 * Protoype for a token address pricing function.
 * @param address Target for price retrieval.
 */
export type PricingFunction = (address: string) => Promise<TokenPrice>;

/**
 * Mass price update request object.
 * Mapping of token address to respective pricing function.
 * @see {PricingFunction}
 */
export interface PriceUpdateRequest {
  [token: string]: PricingFunction;
}

export const priceCache = new NodeCache({ stdTTL: 300, checkperiod: 480 });

/**
 * Update pricing db entry using chain strategy.
 * @param token Target for price update.
 */
export const updatePrice = async (token: Token): Promise<void> => {
  const { address, name } = token;
  const strategy = ChainStrategy.getStrategy(address);
  const tokenPriceData = await strategy.getPrice(address);
  tokenPriceData.name = name;
  tokenPriceData.address = address;
  const tokenPriceSnapshot: TokenPriceSnapshot = {
    ...tokenPriceData,
    updatedAt: Date.now(),
  };
  await saveItem(PRICE_DATA, tokenPriceSnapshot);
};

/**
 * Mass update token pricing db entries.
 * @param tokenConfig Target for price updates.
 */
export const updatePrices = async (tokenConfig: TokenConfig): Promise<void> => {
  await Promise.all(Object.values(tokenConfig).map(async (token) => updatePrice(token)));
};

/**
 * Load token price fromt he pricing database.
 * @param contract Address for the token price being requested.
 * @returns Most recent price data for the requested contract.
 */
export const getPrice = async (contract: string): Promise<TokenPriceSnapshot> => {
  const checksumContract: AttributeValue = ethers.utils.getAddress(contract);
  const params: QueryInput = {
    TableName: PRICE_DATA,
    KeyConditionExpression: 'address = :address',
    ExpressionAttributeValues: {
      ':address': checksumContract,
    },
    Limit: 1,
    ScanIndexForward: false,
  };
  const prices = await getItems<TokenPriceSnapshot>(params);
  if (!prices || prices.length !== 1) {
    throw new NotFound(`No price data stored for ${checksumContract}`);
  }
  return prices[0];
};

/**
 * Retrieve the price data for a given token in USD and ETH.
 * @param contract Token contract address.
 * @returns Most recently updated token pricing data.
 */
export const getTokenPriceData = async (contract: string): Promise<TokenPrice> => {
  const checksumContract = ethers.utils.getAddress(contract);
  const cachedPrice = priceCache.get<TokenPrice>(checksumContract);
  if (!cachedPrice) {
    const priceData = await getPriceData();
    return priceData[checksumContract];
  }
  return cachedPrice;
};

/**
 * Retrieve all chain token prices in both USD and ETH.
 * @returns Most recently updated token pricing data for all tokens.
 */
export const getPriceData = async (): Promise<PriceData> => {
  const priceData: PriceData = {};
  const prices = await Promise.all(Object.keys(protocolTokens).map((key) => getPrice(protocolTokens[key].address)));
  /* eslint-disable @typescript-eslint/no-non-null-assertion */
  prices.forEach((token) => {
    priceData[token.address] = token;
    priceCache.set(token.address, token);
  });
  /* eslint-disable @typescript-eslint/no-non-null-assertion */
  return priceData;
};

/**
 * Retrieve the price data for a given token in USD and ETH.
 * Warning: Not all contracts are supported.
 * If a token is not supported, but available on CoinGecko, use getTokenPrice.
 * @param contract Token contract address.
 * @throws {InternalServerError} Failed price lookup.
 */
export const getContractPrice = async (contract: string): Promise<TokenPrice> => {
  const cachedPrice = priceCache.get<TokenPrice>(contract);
  if (cachedPrice) {
    return cachedPrice;
  }
  const response = await fetch(
    `${COINGECKO_URL}/token_price/ethereum?contract_addresses=${contract}&vs_currencies=usd,eth`,
  );
  if (!response.ok) {
    throw new InternalServerError(`Unable to query ${contract} price`);
  }
  const json = await response.json();
  const contractKey = contract.toLowerCase(); // coingecko return key in lower case
  if (!json[contractKey] || !json[contractKey].usd || !json[contractKey].eth) {
    throw new InternalServerError(`Unable to resolve ${contract} price`);
  }
  const token = getToken(contract);
  const contractPrice: TokenPrice = {
    name: token.name,
    address: token.address,
    usd: json[contractKey].usd,
    eth: json[contractKey].eth,
  };
  priceCache.set(contract, contractPrice);
  return contractPrice;
};

/**
 * Retrieve the price data for a given token in USD and ETH.
 * @param token CoinGecko token name.
 * @throws {InternalServerError} Failed price lookup.
 */
export const getTokenPrice = async (name: string): Promise<TokenPrice> => {
  const cachedPrice = priceCache.get<TokenPrice>(name);
  if (cachedPrice) {
    return cachedPrice;
  }
  const response = await fetch(`${COINGECKO_URL}/price?ids=${name}&vs_currencies=usd,eth`);
  if (!response.ok) {
    throw new InternalServerError(`Unable to query ${name} price`);
  }
  const json = await response.json();
  if (!json[name] || !json[name].usd || !json[name].eth) {
    throw new InternalServerError(`Unable to resolve ${name} price`);
  }
  const token = getTokenByName(name);
  const tokenPrice: TokenPrice = {
    name: token.name,
    address: token.address,
    usd: json[name].usd,
    eth: json[name].eth,
  };
  priceCache.set(name, tokenPrice);
  return tokenPrice;
};

/**
 * Convert a given token price to a defined currency option.
 * @param tokenPrice Pricing data for the given token.
 * @param currency Currency requested from the pricing data.
 * @returns Price value in currency if exists, default to usd if not.
 */
export const inCurrency = (tokenPrice: TokenPrice, currency?: string): number => {
  switch (currency) {
    case 'eth':
      return tokenPrice.eth;
    case 'usd':
    default:
      return tokenPrice.usd;
  }
};

/**
 * Get pricing information for a vault token.
 * @param contract Address for vault token.
 * @returns Pricing data for the given vault token based on the pricePerFullShare.
 */
export const getVaultTokenPrice = async (contract: string): Promise<TokenPrice> => {
  const token = getToken(contract);
  if (token.type !== TokenType.Vault) {
    throw new BadRequest(`${token.name} is not a vault token`);
  }
  if (!token.vaultToken) {
    throw new UnprocessableEntity(`${token.name} vault token missing`);
  }
  const vaultToken = token.vaultToken;
  const targetChain = Chain.getChain(vaultToken.network);
  const [vaultTokenPrice, vaultTokenSnapshot] = await Promise.all([
    getPrice(vaultToken.address),
    getSett(targetChain.graphUrl, token.address),
  ]);
  if (!vaultTokenSnapshot.sett) {
    throw new InternalServerError(`Failed to load ${contract} sett data`);
  }
  vaultTokenPrice.name = token.name;
  vaultTokenPrice.address = token.address;
  const sett = vaultTokenSnapshot.sett;

  let multiplier;
  if (token.address === TOKENS.BDIGG) {
    multiplier = (sett.balance / sett.totalSupply) * 1e9;
  } else {
    multiplier = sett.pricePerFullShare / 1e18;
  }
  vaultTokenPrice.usd *= multiplier;
  vaultTokenPrice.eth *= multiplier;

  return vaultTokenPrice;
};

/**
 * Get pricing information for a wrapper token.
 * Wrapper tokens are tokens on their non-native chain that represent another token
 * in the system. i.e. bDigg on Binance Smart Chain.
 * @param contract Address for wrapper token.
 * @returns Pricing data for the given wrapper token.
 */
export const getWrapperTokenPrice = async (contract: string): Promise<TokenPrice> => {
  const token = getToken(contract);
  if (token.type !== TokenType.Wrapper) {
    throw new BadRequest(`${token.name} is not a wrapper token`);
  }
  if (!token.vaultToken) {
    throw new UnprocessableEntity(`${token.name} vault token missing`);
  }
  const vaultToken = token.vaultToken;
  const allSetts = [];
  for (const chain of loadChains()) {
    allSetts.push(...chain.setts);
  }
  const backingVault = allSetts.find((sett) => ethers.utils.getAddress(sett.depositToken) === vaultToken.address);
  if (!backingVault) {
    throw new UnprocessableEntity(`${token.name} vault token missing`);
  }
  const price = await getPrice(backingVault.settToken);
  if (!price) {
    throw new InternalServerError(`Failed to load ${contract} sett data`);
  }
  return price;
};
