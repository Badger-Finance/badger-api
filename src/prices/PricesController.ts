import { Controller, Get, Inject, QueryParams } from '@tsed/common';
import { ContentType } from '@tsed/schema';
import { Chain } from '../chains/config/chain.config';
import { ChainNetwork } from '../chains/enums/chain-network.enum';
import { PriceSummary } from '../tokens/interfaces/token-price.interface';
import { PricesService } from './PricesService';

@Controller('/prices')
export class PriceController {
  @Inject()
  pricesService!: PricesService;

  @ContentType('json')
  @Get('/')
  async listPrices(
    @QueryParams('chain') chain?: ChainNetwork,
    @QueryParams('currency') currency?: string,
  ): Promise<PriceSummary> {
    return this.pricesService.getPriceSummary(Chain.getChain(chain), currency);
  }
}
