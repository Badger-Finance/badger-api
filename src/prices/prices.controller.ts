import { Currency, Network } from '@badger-dao/sdk';
import { Controller, Inject } from '@tsed/di';
import { UseCache } from '@tsed/platform-cache';
import { QueryParams } from '@tsed/platform-params';
import { ContentType, Description, Get, Hidden, Returns, Summary } from '@tsed/schema';

import { Chain } from '../chains/config/chain.config';
import { PriceSnapshots } from '../tokens/interfaces/price-snapshots.interface';
import { PriceSummary } from '../tokens/interfaces/price-summary.interface';
import { PriceSummaryModel } from '../tokens/interfaces/price-summary-model.interface';
import { PricesService } from './prices.service';

@Controller('/prices')
export class PricesController {
  @Inject()
  pricesService!: PricesService;

  @Get('')
  @UseCache()
  @ContentType('json')
  @Summary('Get a summary of token prices related to the Badger Protocol')
  @Description('Return a map of checksum contract address to the currency value of the token')
  @Returns(200, PriceSummaryModel)
  async listPrices(
    @QueryParams('tokens') tokens?: string,
    @QueryParams('chain') chain?: Network,
    @QueryParams('currency') currency?: Currency
  ): Promise<PriceSummary> {
    return this.pricesService.getPriceSummary(
      tokens?.split(',') ?? Object.keys(Chain.getChain(chain).tokens),
      currency
    );
  }

  @Get('/snapshots')
  @Hidden()
  @UseCache()
  @ContentType('json')
  async getPriceSnapshots(
    @QueryParams('tokens') tokens: string,
    @QueryParams('timestamps') timestamps: string,
    @QueryParams('currency') currency?: Currency
  ): Promise<PriceSnapshots> {
    return this.pricesService.getPriceSnapshots(
      tokens.split(','),
      timestamps.split(',').map((t) => Number(t)),
      currency
    );
  }
}
