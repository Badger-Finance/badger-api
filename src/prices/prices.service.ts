import { Currency } from '@badger-dao/sdk';
import { Service } from '@tsed/common';

import { Chain } from '../chains/config/chain.config';
import { PriceSummary } from '../tokens/interfaces/price-summary.interface';
import { convert, getPrice } from './prices.utils';

/**
 * API price oracle service. Uses CoinGecko as a source of truth for most
 * tokens when possible, and TheGraph for AMM pairs when not available
 * via CG. Prices are cached for 5 minutes at a time, but may live up to 8.
 */
@Service()
export class PricesService {
  async getPriceSummary(chain: Chain, currency?: Currency): Promise<PriceSummary> {
    const prices = await Promise.all(Object.keys(chain.tokens).map(async (token) => getPrice(token)));
    const entries = await Promise.all(
      prices.map(async (tokenPrice) => {
        const convertedPrice = await convert(tokenPrice.price, currency);
        return [tokenPrice.address, convertedPrice];
      }),
    );
    return Object.fromEntries(entries);
  }
}
