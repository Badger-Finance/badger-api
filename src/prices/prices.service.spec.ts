import { Currency, ONE_DAY_MS } from '@badger-dao/sdk';
import { PlatformTest } from '@tsed/common';

import { TEST_ADDR, TEST_CURRENT_TIMESTAMP, TEST_TOKEN } from '../test/constants';
import { setupMockChain } from '../test/mocks.utils';
import { PricesService } from './prices.service';

describe('prices.service', () => {
  let service: PricesService;

  beforeEach(async () => {
    await PlatformTest.create();
    service = PlatformTest.get<PricesService>(PricesService);
    setupMockChain();
  });

  afterEach(PlatformTest.reset);

  describe('getPriceSummary', () => {
    it('returns a price summary for the requested chains tokens', async () => {
      const results = await service.getPriceSummary([TEST_TOKEN, TEST_ADDR]);
      expect(results).toMatchSnapshot();
    });
  });

  describe('getPriceSnapshots', () => {
    describe('request with no tokens', () => {
      it('returns an empty map', async () => {
        const results = await service.getPriceSnapshots([], [TEST_CURRENT_TIMESTAMP]);
        expect(results).toMatchObject({});
      });
    });

    describe('request with no timestamps', () => {
      it('returns an empty map', async () => {
        const results = await service.getPriceSnapshots([TEST_TOKEN, TEST_ADDR], []);
        expect(results).toMatchObject({
          [TEST_TOKEN]: {},
          [TEST_ADDR]: {},
        });
      });
    });

    describe('proper full request with no currency', () => {
      it('returns an empty map', async () => {
        const results = await service.getPriceSnapshots(
          [TEST_TOKEN, TEST_ADDR],
          [TEST_CURRENT_TIMESTAMP, TEST_CURRENT_TIMESTAMP - ONE_DAY_MS],
        );
        expect(results).toMatchSnapshot();
      });
    });

    describe('proper full request with currency', () => {
      it('returns an empty map', async () => {
        const results = await service.getPriceSnapshots(
          [TEST_TOKEN, TEST_ADDR],
          [TEST_CURRENT_TIMESTAMP, TEST_CURRENT_TIMESTAMP - ONE_DAY_MS],
          Currency.AVAX,
        );
        expect(results).toMatchSnapshot();
      });
    });
  });
});
