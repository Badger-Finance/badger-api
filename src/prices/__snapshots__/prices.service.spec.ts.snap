// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`leaderboards.service getPriceSummary returns a price summary for the requested chains tokens 1`] = `
Object {
  "0x0316EB71485b0Ab14103307bf65a021042c6d380": 130.66666666666666,
  "0x04c90C198b2eFF55716079bc06d7CCc4aa4d7512": 202.66666666666666,
  "0x075b1bb99792c9E1041bA13afEf80C91a1e70fB3": 312,
  "0x110492b31c59716AC47337E616804E3E3AdC0b4a": 725.3333333333334,
  "0x15cBC4ac1e81c97667780fE6DAdeDd04a6EEB47B": 928,
  "0x17d8CBB6Bce8cEE970a4027d1198F6700A7a6c24": 1016,
  "0x1862A18181346EBd9EdAf800804f89190DeF24a5": 1040,
  "0x18d98D452072Ac2EB7b74ce3DB723374360539f1": 1058.6666666666667,
  "0x19D97D8fA813EE2f51aD4B4e04EA08bAf4DFfC28": 1101.3333333333333,
  "0x19E4d89e0cB807ea21B8CEF02df5eAA99A110dA5": 1104,
  "0x20c36f062a31865bED8a5B1e512D9a1A20AA333A": 1397.3333333333333,
  "0x2260FAC5E5542a773Aa44fBCfeDf7C193bc2C599": 1466.6666666666667,
  "0x235c9e24D3FB2FAFd58a2E49D454Fdcd2DBf7FF1": 1506.6666666666667,
  "0x26B8efa69603537AC8ab55768b6740b67664D518": 1650.6666666666667,
  "0x27E98fC7d05f54E544d16F58C194C2D7ba71e3B5": 1701.3333333333333,
  "0x2B5455aac8d64C14786c3a29858E43b5945819C0": 1848,
  "0x2fE94ea3d5d4a175184081439753DE15AeF9d614": 2042.6666666666667,
  "0x3472A5A71965499acd81997a54BBA8D852C6E53d": 2237.3333333333335,
  "0x3c9d6c1C73b31c837832c72E04D3152f051fc1A9": 2584,
  "0x410e3E86ef427e30B9235497143881f717d93c2A": 2773.3333333333335,
  "0x48c59199Da51B7E30Ea200a74Ea07974e62C4bA7": 3104,
  "0x49849C98ae39Fff122806C06791Fa73784FB3675": 3136,
  "0x4b92d19c11435614CD49Af1b589001b7c08cD4D5": 3224,
  "0x4e3FBD56CD56c3e72c1403e103b45Db9da5B9D2B": 3336,
  "0x5228a22e72ccC52d415EcFd199F99D0665E7733b": 3504,
  "0x53C8E199eb2Cb7c01543C137078a038937a68E40": 3573.3333333333335,
  "0x55912D0Cf83B75c492E761932ABc4DB4a5CB1b17": 3650.6666666666665,
  "0x599D92B453C010b1050d31C364f6ee17E819f193": 3821.3333333333335,
  "0x5Dce29e92b1b939F8E8C60DcF15BDE82A85be4a9": 4000,
  "0x5a6A4D54456819380173272A5E8E9B9904BdF41B": 3856,
  "0x62B9c7356A2Dc64a1969e19C23e4f579F9810Aa7": 4210.666666666667,
  "0x64eda51d3Ad40D56b9dFc5554E06F94e1Dd786Fd": 4304,
  "0x68c269B60c58c4ed50c63B217BA0EC7f8a371920": 4469.333333333333,
  "0x6B175474E89094C44Da98b954EedeAC495271d0F": 4568,
  "0x6B3595068778DD592e39A122f4f5a5cF09C90fE2": 4573.333333333333,
  "0x6aF7377b5009d7d154F36FE9e235aE1DA27Aea22": 4562.666666666667,
  "0x6c3F90f043a72FA612cbac8115EE7e52BDe6E490": 4616,
  "0x6dEf55d2e18486B9dDfaA075bc4e4EE0B28c1545": 4688,
  "0x758A43EE2BFf8230eeb784879CdcFF4828F2544D": 5013.333333333333,
  "0x798D1bE841a82a273720CE31c822C61a67a601C3": 5184,
  "0x7e7E112A68d8D2E221E11047a72fFC1065c38e1a": 5394.666666666667,
  "0x8064d9Ae6cDf087b1bcd5BDf3531bD5d8C537a68": 5477.333333333333,
  "0x853d955aCEf822Db058eb8505911ED77F175b99e": 5682.666666666667,
  "0x85Eee30c52B0b379b046Fb0F85F4f3Dc3009aFEC": 5712,
  "0x8751D4196027d4e6DA63716fA7786B5174F04C15": 5773.333333333333,
  "0x8798249c2E607446EfB7Ad49eC89dD1865Ff4272": 5784,
  "0x88128580ACdD9c04Ce47AFcE196875747bF2A9f6": 5805.333333333333,
  "0x89Ab32156e46F46D02ade3FEcbe5Fc4243B9AAeD": 5872,
  "0x8a8FFec8f4A0C8c9585Da95D9D97e8Cd6de273DE": 5909.333333333333,
  "0x8c76970747afd5398e958bDfadA4cf0B9FcA16c4": 5992,
  "0x8dAEBADE922dF735c38C80C7eBD708Af50815fAa": 6042.666666666667,
  "0x937B8E917d0F36eDEBBA8E459C5FB16F3b315551": 6290.666666666667,
  "0x945Facb997494CC2570096c74b5F66A3507330a1": 6328,
  "0x952F4Ac36EF204a28800AA1c1586C5261B600894": 6362.666666666667,
  "0x99D8a9C45b2ecA8864373A26D1459e3Dff1e17F3": 6562.666666666667,
  "0x99F39D495C6A5237f43602f3Ab5F49786E46c9B0": 6568,
  "0x9BE89D2a4cd102D8Fecc6BF9dA793be995C22541": 6650.666666666667,
  "0x9a13867048e01c663ce8Ce2fE0cDAE69Ff9F35E3": 6573.333333333333,
  "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48": 6856,
  "0xAf5A1DECfa95BAF63E0084a35c62592B774A2A87": 7480,
  "0xBC19712FEB3a26080eBf6f2F7849b417FdD792CA": 8024,
  "0xBE08Ef12e4a553666291E9fFC24fCCFd354F2Dd2": 8106.666666666667,
  "0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2": 8197.333333333334,
  "0xC17078FDd324CC473F8175Dc5290fae5f2E84714": 8253.333333333334,
  "0xCEfF51756c56CeFFCA006cD410B03FFC46dd3a58": 8829.333333333334,
  "0xD533a949740bb3306d119CC777fa900bA034cd52": 9096,
  "0xDE5331AC4B3630f94853Ff322B66407e0D6331E8": 9485.333333333334,
  "0xE143aA25Eec81B4Fc952b38b6Bca8D2395481377": 9610.666666666666,
  "0xE86204c4eDDd2f70eE00EAd6805f917671F56c52": 9914.666666666666,
  "0xEB4C2781e4ebA804CE9a9803C67d0893436bB27D": 10037.333333333334,
  "0xFbdCA68601f835b27790D98bbb8eC7f05FDEaA9B": 10744,
  "0xa0246c9032bC3A600820415aE600c6388619A14D": 6832,
  "0xa3BeD4E1c75D00fa6f4E5E6922DB7261B5E9AcD2": 6984,
  "0xaE96fF08771a109dc6650a1BdCa62F2d558E40af": 7448,
  "0xb19059ebb43466C323583928285a49f558E572Fd": 7576,
  "0xb9D076fDe463dbc9f915E5392F807315Bf940334": 7928,
  "0xc4AD29ba4B3c580e6D59105FFf484999997675Ff": 8389.333333333334,
  "0xc4E15973E6fF2A35cC804c2CF9D2a1b817a8b40F": 8400,
  "0xcA3d75aC011BF5aD07a98d02f18225F9bD9A6BDF": 8626.666666666666,
  "0xcD7989894bc033581532D2cd88Da5db0A4b12859": 8765.333333333334,
  "0xd04c48A53c111300aD41190D63681ed3dAd998eC": 8885.333333333334,
  "0xd632f22692FaC7611d2AA1C0D552930D43CAEd3B": 9138.666666666666,
  "0xdAC17F958D2ee523a2206206994597C13D831ec7": 9333.333333333334,
  "0xf349c0faA80fC1870306Ac093f75934078e28991": 10378.666666666666,
  "0xfE18be6b3Bd88A2D2A7f928d00292E7a9963CfC6": 10840,
  "0xfd05D3C7fe2924020620A8bE4961bBaA747e6305": 10794.666666666666,
}
`;
